### DISCUSSION TITLE
Customize validator and generator using workspace.ConfigurationProvider

### DISCUSSION THREAD

**Discussion by borisbrodski** [eclipse-langium/langium#1651] [Q&A] [answered]
My DSL needs some project dependent configuration, like for example:

* name of an abstract class to extend my generated java class from
* some limits to stuff

The settings should work the same: in LS (VS Code/Eclipse/....) and in CLI generator.

I looked into using `workspace.ConfigurationProvider` for this purpose. The settings could be declared and described in `package.json` and the values places then into `.vscode/settings.json`.

But `settings.json` doesn't get loaded while generating code with CLI. Other IDEs would probably also not be capable of reading 
`settings.json` and providing LS with my custom configuration.

Is there some working solution/example of how to deal with settings in Langium DSLs?

Or sould I implement my own configuration, which is independend from LS and VSCode features?

Thank you!

**Comment by msujew** [comment, marked as answer]
Hey @borisbrodski,

the configuration you're talking about is essentially a compiler configuration. Most programming languages (like TypeScript's `tsconfig.json`) feature a separate configuration format for that. The CLI and the language server would both read that file to determine the compiler options.

In general, you shouldn't use compiler options as language server configuration items. As you've already seen, reading those in a CLI is effectively impossible and there's even one more important argument: The language server provides language services to an IDE. Therefore, unless absolutely necessary, only language service related configurations should appear as configuration items (i.e. stuff like whether certain code lenses should appear, etc.).

**Comment by borisbrodski** [comment]
Hey Mark,

thank you for quick response!

This makes a lot of sense. How would you implement loading `mydsl-config.json` files?

* Create a custom `WorkspaceManager`, that loads `my-config.json` configuration files overwriting `DefaultWorkspaceManager.loadAdditionalDocuments()` method and add these as `LangiumDocument`s?

I would rather write an additional service like `MySettingsProvider`, but I found no way to access the workspace paths used to initialize `WorkspaceManager`.

I have following in my `main.ts`:

```
    const workspaceFolder: WorkspaceFolder = {
        name: 'MyDSL Workspace',
        uri: URI.file(path.resolve(workspaceDir)).toString()
    };
    await workspaceManager.initializeWorkspace([workspaceFolder]);
```

**Reply by msujew** [reply]
It's likely much more reasonable to have a `MySettingsProvider` that works with the `json` file directly. Stuffing the data into a `LangiumDocument` will be quite difficult, as those are reserved for documents of your DSL.

> I found no way to access the workspace paths used to initialize WorkspaceManager.

You can listen on the initialization of the language server, see 

https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/lsp/language-server.ts#L46

The `params` contain the workspace folders of the language client. Alternatively, you can create your own `WorkspaceManager` that initializes the `MySettingsProvider` with the provided workspace folders.
